src/client/dbif.cc

```diff
@@ -0,0 +1,953 @@
+/*
+ * Copyright 2017 CodiLime
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <memory>
+#include <vector>
+
+#include <QSharedPointer>
+
+#include "data/types.h"
+#include "network/msgpackwrapper.h"
+#include "client/dbif.h"
+
+namespace veles {
+namespace client {
+
+/*****************************************************************************/
+/* NCObjectHandle */
+/*****************************************************************************/
+
+NCObjectHandle::NCObjectHandle(NCWrapper* nc, data::NodeID id,
+    dbif::ObjectType type) : nc_(nc), id_(id), type_(type) {
+}
+
+data::NodeID NCObjectHandle::id() {
+  return id_;
+}
+
+dbif::InfoPromise* NCObjectHandle::getInfo(dbif::PInfoRequest req) {
+  return nc_->getInfo(req, id_);
+}
+
+dbif::InfoPromise* NCObjectHandle::subInfo(dbif::PInfoRequest req) {
+  return nc_->subInfo(req, id_);
+}
+
+dbif::MethodResultPromise* NCObjectHandle::runMethod(
+    dbif::PMethodRequest req) {
+  return nc_->runMethod(req, id_);
+}
+
+dbif::ObjectType NCObjectHandle::type() const {
+  return type_;
+}
+
+bool NCObjectHandle::operator==(NCObjectHandle& other) {
+  return id_ == other.id_;
+}
+
+/*****************************************************************************/
+/* NCWrapper */
+/*****************************************************************************/
+
+NCWrapper::NCWrapper(NetworkClient* network_client, QObject* parent)
+    : QObject(parent), nc_(network_client), detailed_debug_info_(true) {
+  if (nc_) {
+    message_handlers_["get_list_reply"]
+        = &NCWrapper::handleGetListReplyMessage;
+    message_handlers_["request_ack"]
+        = &NCWrapper::handleRequestAckMessage;
+    message_handlers_["get_reply"]
+        = &NCWrapper::handleGetReplyMessage;
+    message_handlers_["get_data_reply"]
+        = &NCWrapper::handleGetDataReplyMessage;
+    message_handlers_["get_bindata_reply"]
+        = &NCWrapper::handleGetBinDataReplyMessage;
+    message_handlers_["query_error"]
+        = &NCWrapper::handleQueryErrorMessage;
+
+    connect(nc_, &NetworkClient::connectionStatusChanged,
+        this, &NCWrapper::updateConnectionStatus);
+    connect(nc_, &NetworkClient::messageReceived,
+        this, &NCWrapper::messageReceived);
+  }
+}
+
+dbif::ObjectType NCWrapper::typeFromTags(
+      std::shared_ptr<std::unordered_set<std::shared_ptr<std::string>>> tags) {
```

Comment:
Eh, now that I see it used, I see how it sucks. We either need to remove pointers from strings or provide more template parameters to this template to make it usable in any way - I'll prepare a patch with removing pointers for now.